File: ProjectEuler/051-100/S70.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ProjectEuler
{
    public class S70 : ISolution
    {
        public string GetAnswer()
        {
            var primesForTesting = Utils.GenPrimes(10000000);
            int[] primes = Utils.GenPrimes(1000).Where(l => l != 0).Select(Convert.ToInt32).ToArray();

            int n = 0;
            double minRatio = double.MaxValue;

            List<int> numbers = new List<int>();
            for (int i = 1000000; i < 10000000; i++)
            {
                if (Meet(i, primes))
                {
                    numbers.Add(i);
                }
            }

            foreach (var i in numbers)
            {
                if (primesForTesting[i] != 0)
                {
                    continue;
                }

                int count = Utils.GetCoprimeCount(i);
                if (((double)i) / count < minRatio)
                {
                    if (Utils.IsPermutation(i, count))
                    {
                        minRatio = ((double)i) / count;
                        n = i;
                    }
                }
            }

            return n.ToString();
        }

        private static bool Meet(int i, int[] primes)
        {
            foreach (var prime in primes)
            {
                if (i % prime == 0)
                {
                    return false;
                }
            }

            return true;
        }
    }
}

File: src/051-100/S70.cc
#include "S70.h"
#include "../Utils.h"
#include <algorithm>
#include <cmath>
#include <string>
#include <vector>

using namespace std;

namespace ProjectEuler
{
bool IsPermutation(int64_t a, int64_t b)
{
	string sa = to_string(a);
	string sb = to_string(b);
	if (sa.length() != sb.length())
	{
		return false;
	}
	sort(sa.begin(), sa.end());
	sort(sb.begin(), sb.end());
	return sa == sb;
}

std::string S70::GetAnswer()
{
	int64_t limit = 10000000;
	double minRatio = 100.0;
	int64_t result = 0;

	auto primes = Utils::GenPrimes(5000);
	std::vector<int> rangePrimes;
	for (int64_t p : primes)
	{
		if (p > 2000)
		{
			rangePrimes.push_back(p);
		}
	}

	for (size_t i = 0; i < rangePrimes.size(); i++)
	{
		for (size_t j = i; j < rangePrimes.size(); j++)
		{
			int64_t n = (int64_t)rangePrimes[i] * rangePrimes[j];
			if (n > limit)
			{
				break;
			}

			int64_t phi = (rangePrimes[i] - 1) * (rangePrimes[j] - 1);

			double ratio = (double)n / phi;
			if (ratio < minRatio)
			{
				if (IsPermutation(n, phi))
				{
					minRatio = ratio;
					result = n;
				}
			}
		}
	}

	return std::to_string(result);
}
} // namespace ProjectEuler
